### PR TITLE
Clamp the zoneviewwidget's topleft point to the scene view area; fix #754

### LINKED ISSUE #754: Can't close "view deck" dialog anymore
On  small screen (13" mac book pro). Trying to make the window wider, the dialog moves accordingly keeping the "X" button outside the screen.
Maybe an hotkey like "ESC" could help.

schermata 2015-02-17 alle 23 31 00

### PR DESCRIPTION
**Pull request by ctrlaltca**
...4

### CODE REVIEW

**Comment by Daenyth** [comment]
I think you can reword that body to be 1-2 lines with `max` calls. I'll work it out

File: cockatrice/src/zoneviewwidget.cpp [hunk 1]
```diff
@@ -185,6 +185,24 @@ void ZoneViewWidget::paint(QPainter *painter, const QStyleOptionGraphicsItem *op
...
 void ZoneViewWidget::moveWidget(QPointF scenePos)
 {
+    if(scenePos.x() < 0)
+    {
+        scenePos.setX(0);
+    } else {
+        qreal maxw = scene()->sceneRect().width() - 100;
+        if(scenePos.x() > maxw)
+            scenePos.setX(maxw);
+    }
```
Comment (Daenyth) [inline cockatrice/src/zoneviewwidget.cpp]:
```
qreal maxw = scene()->sceenRect().width() - 100; // where does 100 come from?
scenePos.setX( std::min(maxw, std::max(scenePos.x(), 0)));
```
And similarly for Y

**Comment by ZeldaZach** [comment]
I don't know how to test this exactly as nothing looks different for me when viewing zones in games...

**Comment by Daenyth** [comment]
Try with a deck of 200 lands and pile view on

**Comment by ctrlaltca** [comment]
Try to reveal your deck and then drag the zoneview around; you should notice that you won't be able to move its titlebar completely outside the battlefield.

**Comment by ctrlaltca** [comment]
Let me try to explain this better with my awesome paint skillz.
When a player reveals his library, a "zoneviewwidget" appears, hiding the battlefield. Its default position is (50px, 50px).
default_pos

You can then drag it around using the mouse; if by accident you move its titlebar outside the battlefield (gameviewscene) area, you won't be able to move it back anymore:
out_of_scene

This PR tries to address this issue, forcing the titlebar (or at least a small part of it) to always stay visible on the battefield. I basically clamp the zoneviewwidget's topleft point between the (0px, 0px) position:
topleft

and the (scene width -100px, scene height - 100px) position:
bottomright

The "100px" number that i choose has no meaning, is completely artificial, the same way as the (50px, 50px) initial position of the zoneviewwidget.

File: cockatrice/src/zoneviewwidget.cpp (hunk 1 above)
Comment (ctrlaltca) [inline cockatrice/src/zoneviewwidget.cpp]: Sounds like less lines of code, but more cycles of cpu:
 * if the value is already ok (99% of the time), you'll just call setX() with the very same value;
 * if the value is less than 0,  std::max() will set it to 0; in this case we already know that 0 is less than maxw, so there's no need to std::min();
 * the value of maxw is always calculated, even if it's not really needed;

I agree that your code looks nicer, but given that this method is called on a mousemove event that can be triggered tens of times per second, i'd prefer to keep it faster.

Comment (Daenyth) [inline cockatrice/src/zoneviewwidget.cpp]: That's fair, I wasn't considering the frequency that it's triggered.

Keep it as it. And if you notice me adding code like that in often-called places please point it out during review..

**Pull request merged**